crates/store/src/db/models/queries/accounts.rs

```diff
@@ -1010,3 +1205,235 @@ pub(crate) struct AccountStorageMapRowInsert {
     pub(crate) value: Vec<u8>,
     pub(crate) is_latest: bool,
 }
+
+/// Queries vault assets (key, value) pairs at a specific block
+pub(crate) fn select_account_vault_at_block(
+    conn: &mut SqliteConnection,
+    account_id: AccountId,
+    block_num: BlockNumber,
+) -> Result<Vec<(Word, Word)>, DatabaseError> {
+    use schema::account_vault_assets as t;
+
+    // Check if the requested block exists (returns error if not)
+    block_exists(conn, block_num)?;
+
+    let account_id_bytes = account_id.to_bytes();
+    let block_num_sql = i64::from(block_num.as_u32());
+    let raw: Vec<(Vec<u8>, Option<Vec<u8>>)> = SelectDsl::select(
+        t::table
+            .filter(t::account_id.eq(&account_id_bytes))
+            .filter(t::block_num.le(block_num_sql))
+            .order(t::block_num.desc())
+            .limit(1),
+        (t::vault_key, t::asset),
+    )
+    .load(conn)?;
```

Comment:
Would this not always return just 1 entry (because of `LIMIT 1`)? I think we may need to get all entries with `block_num` smaller than or equal to the passed in block number and then perform filtering manually - but maybe there is a better way to do it.